Support custom resolvers.

Aeyk opened the pull request:
**Changes**
Allow additional scanners to read metadata from the file system structure of folders during a library scan. This should help support [custom resolvers](https://github.com/jellyfin/jellyfin/discussions/5732) as plugins. 

**Issues**
[5732](https://github.com/jellyfin/jellyfin/discussions/5732)

Aeyk committed:
Support additional metadata scanners.

IDisposable commented on MediaBrowser.Controller/Resolvers/Scanner.cs:
```diff
@@ -0,0 +1,38 @@
+
+        protected BaseItem ApplyMetadata(BaseItem t)
+        {
+            throw new NotImplementedException();
```
This would prevent safe use of calling `t = base.ApplyMetadata(t);` in multiple-layer derived resolver classes. Do we explicitly want to discourage that sort of coding?

Perhaps better to just make it this 
```c#
return t;
```

JPVenson replied:
yes.

this is the base function/only function that actually does something here. There is no other reason to implement this base class, so why not just make this function abstract itself?

```suggestion
            protected abstract BaseItem ApplyMetadata(BaseItem t);
```

IDisposable commented on MediaBrowser.Providers/Manager/ProviderManager.cs:
```diff
@@ -611,10 +619,10 @@ private void AddMetadataPlugins<T>(List<MetadataPlugin> list, T item, LibraryOpt
 
             // Locals
-            list.AddRange(providers.Where(i => i is ILocalMetadataProvider).Select(i => new MetadataPlugin
+            list.AddRange(providers.Where(i => i is IMetadataProvider && i is not IRemoteMetadataProvider && i is not ILocalMetadataProvider && i is not ICustomMetadataProvider).Select(i => new MetadataPlugin
```
Doesn't this kill off the `ILocalMetadataProvider` derived classes that the old code explicitly included?

Also seems very fragile to test for a (limited) set of `is not` instead of explicitly listing the marker interfaces we _want_ to include.

IDisposable commented on MediaBrowser.Providers/Manager/ProviderManager.cs:
```diff
@@ -591,6 +591,14 @@ private MetadataPluginSummary GetPluginSummary<T>()
+            {
+                Name = i?.Name,
+                Type = MetadataPluginType.MetadataResolver
+            }));
```
... or here add a  `.Where(i => i.Name != null);` (inside last parathesis)?

IDisposable commented on MediaBrowser.Providers/Manager/ProviderManager.cs:
```diff
@@ -591,6 +591,14 @@ private MetadataPluginSummary GetPluginSummary<T>()
             }));
 
+            // Filesystem to metadata resolvers
+            var metadataResolverProviders = summary.Plugins.Select(p => p.Type == MetadataPluginType.MetadataResolver && p.Name != null ? p : null);
```
Do we want to tack on a `.Where(i != null)` here?

JPVenson replied:
how _could_ they ever be null?

IDisposable replied:
I was referring to the adding onto the block `.Select(p => p.Type == MetadataPluginType.MetadataResolver && p.Name != null ? p : null)`, where what comes out of the `.Select` could be null if the `.Type` is not `MetadataPluginType.MetadataResolver` OR if it has a `null` setting on the `.Name`... in those cases we would get a `null` out of the ternary... so maybe we should filter/skip those?

`.Select(p => p.Type == MetadataPluginType.MetadataResolver && p.Name != null ? p : null).Where(n => n != null);`

Aeyk committed:
bugfix: make function abstract

This should allow safely calling `t = base.ApplyMetadata(t)` in derived classes.

Co-authored-by: JPVenson <ger-delta-07@hotmail.de>
